File: plugins/togetherai-skills/skills/together-images/scripts/generate_image.py
#!/usr/bin/env python3
"""
Together AI Image Generation -- Text-to-Image and FLUX.2 (v2 SDK)

Generate images from text prompts, save locally, create variations,
and use FLUX.2 reference images.

Usage:
    python generate_image.py

Requires:
    uv pip install "together>=2.0.0"
    export TOGETHER_API_KEY=your_key
"""

import base64
from together import Together

client = Together()


def generate_image_url(
    prompt: str,
    model: str = "black-forest-labs/FLUX.2-dev",
    width: int = 1024,
    height: int = 1024,
    steps: int = 20,
    n: int = 1,
    seed: int | None = None,
) -> list[str]:
    """Generate image(s) and return URL(s)."""
    kwargs: dict = dict(
        model=model,
        prompt=prompt,
        width=width,
        height=height,
        steps=steps,
        n=n,
    )
    if seed is not None:
        kwargs["seed"] = seed

    response = client.images.generate(**kwargs)
    urls = [img.url for img in response.data]
    for i, url in enumerate(urls):
        print(f"  Image {i}: {url}")
    return urls


def generate_and_save(
    prompt: str,
    output_path: str = "output.png",
    model: str = "black-forest-labs/FLUX.2-dev",
    width: int = 1024,
    height: int = 1024,
    steps: int = 20,
) -> str:
    """Generate an image and save it locally via base64."""
    response = client.images.generate(
        model=model,
        prompt=prompt,
        width=width,
        height=height,
        steps=steps,
        n=1,
        response_format="base64",
    )
    image_data = base64.b64decode(response.data[0].b64_json)

    with open(output_path, "wb") as f:
        f.write(image_data)

    print(f"  Saved to {output_path} ({len(image_data)} bytes)")
    return output_path


def generate_flux2(
    prompt: str,
    model: str = "black-forest-labs/FLUX.2-pro",
    width: int = 1024,
    height: int = 768,
    reference_images: list[str] | None = None,
    prompt_upsampling: bool = True,
    output_format: str = "png",
) -> str:
    """Generate with FLUX.2 features (prompt upsampling, reference images)."""
    kwargs: dict = dict(
        model=model,
        prompt=prompt,
        width=width,
        height=height,
        prompt_upsampling=prompt_upsampling,
        output_format=output_format,
    )
    if reference_images:
        kwargs["reference_images"] = reference_images

    response = client.images.generate(**kwargs)
    url = response.data[0].url
    print(f"  FLUX.2 image: {url}")
    return url


if __name__ == "__main__":
    # --- Example 1: Basic text-to-image ---
    print("=== Basic Generation ===")
    generate_image_url(
        prompt="A serene mountain landscape at sunset, digital art",
        steps=20,
    )

    # --- Example 2: Save locally ---
    print("\n=== Save to File ===")
    generate_and_save(
        prompt="A futuristic city skyline with flying cars",
        output_path="city.png",
        steps=20,
    )

    # --- Example 3: Multiple variations ---
    print("\n=== 3 Variations ===")
    generate_image_url(
        prompt="A cute robot reading a book",
        n=3,
        steps=20,
    )

    # --- Example 4: Reproducible with seed ---
    print("\n=== Reproducible (seed=42) ===")
    generate_image_url(
        prompt="Abstract geometric pattern in blue and gold",
        seed=42,
        steps=20,
    )

    # --- Example 5: FLUX.2 with prompt upsampling ---
    print("\n=== FLUX.2 Pro ===")
    generate_flux2(
        prompt="A mountain landscape at sunset with golden light reflecting on a calm lake",
    )

    # --- Example 6: FLUX.2 with reference image ---
    # print("\n=== FLUX.2 Reference Image ===")
    # generate_flux2(
    #     prompt="Replace the color of the car to blue",
    #     reference_images=["https://images.pexels.com/photos/3729464/pexels-photo-3729464.jpeg"],
    # )

File: plugins/togetherai-skills/skills/together-images/scripts/generate_image.ts
#!/usr/bin/env -S npx tsx
/**
 * Together AI Image Generation -- Text-to-Image, Editing, FLUX.2
 *
 * Generate images from text, edit with Kontext, use FLUX.2 features.
 *
 * Usage:
 *   npx tsx generate_image.ts
 *
 * Requires:
 *   npm install together-ai
 *   export TOGETHER_API_KEY=your_key
 */

import Together from "together-ai";

const client = new Together({
  apiKey: process.env.TOGETHER_API_KEY,
});

async function basicGeneration(): Promise<void> {
  console.log("=== Basic Generation (FLUX.2 Dev) ===");
  const response = await client.images.generate({
    model: "black-forest-labs/FLUX.2-dev",
    prompt: "A serene mountain landscape at sunset with a lake reflection",
    steps: 20,
  });
  console.log(`  Image URL: ${response.data[0].url}`);
}

async function flux2Generation(): Promise<void> {
  console.log("\n=== FLUX.2 Pro ===");
  const response = await client.images.generate({
    model: "black-forest-labs/FLUX.2-pro",
    prompt: "A mountain landscape at sunset with golden light reflecting on a calm lake",
    width: 1024,
    height: 768,
    prompt_upsampling: true,
    output_format: "png",
  });
  console.log(`  Image URL: ${response.data[0].url}`);
}

async function kontextEditing(): Promise<void> {
  console.log("\n=== Kontext Image Editing ===");
  const response = await client.images.generate({
    model: "black-forest-labs/FLUX.1-kontext-pro",
    prompt: "Transform this into a watercolor painting",
    image_url: "https://cdn.pixabay.com/photo/2020/05/20/08/27/cat-5195431_1280.jpg",
    width: 1024,
    height: 768,
    steps: 28,
  });
  console.log(`  Edited image: ${response.data[0].url}`);
}

async function multipleVariations(): Promise<void> {
  console.log("\n=== Multiple Variations ===");
  const response = await client.images.generate({
    model: "black-forest-labs/FLUX.2-dev",
    prompt: "A cute robot assistant helping in a modern office",
    n: 4,
    steps: 20,
  });
  for (let i = 0; i < response.data.length; i++) {
    console.log(`  Variation ${i + 1}: ${response.data[i].url}`);
  }
}

async function base64Response(): Promise<void> {
  console.log("\n=== Base64 Response ===");
  const response = await client.images.generate({
    model: "black-forest-labs/FLUX.2-dev",
    prompt: "A cat in outer space",
    steps: 20,
    response_format: "base64",
  });
  const data = response.data[0].b64_json ?? "";
  console.log(`  Base64 length: ${data.length} chars`);
}

async function main(): Promise<void> {
  await basicGeneration();
  await flux2Generation();
  await kontextEditing();
  await multipleVariations();
  await base64Response();
}

main();
